src/main/java/com/metamx/common/FileUtils.java

```diff
@@ -0,0 +1,302 @@
+   */
+  public static FileCopyResult retryCopy(
+      final ByteSource byteSource,
+      final File outFile,
+      final Predicate<Throwable> shouldRetry,
+      final int maxAttempts
+  )
+  {
+    try {
+      return RetryUtils.retry(
+          new Callable<FileCopyResult>()
+          {
+            @Override
+            public FileCopyResult call() throws Exception
+            {
+              try (final InputStream inputStream = byteSource.openStream()) {
+                if (outFile.exists()) {
+                  if (!outFile.delete()) {
+                    throw new IAE("Could not remove file at [%s]", outFile.getAbsolutePath());
+                  }
+                }
+
+                final MessageDigest digest = StreamUtils.copyToFileWithInputDigest(inputStream, outFile, "MD5");
+                final FileCopyResult result = FileCopyResult.buildResult(outFile.toPath(), "MD5");
+                if (!result.digestEquals(digest)) {
+                  throw new MD5MismatchException(
+                      "File at [%s] did not match MD5 after copy",
+                      outFile.getAbsolutePath()
+                  );
+                }
+                return result;
+              }
+            }
+          },
+          shouldRetry,
+          maxAttempts
+      );
+    }
+    catch (Exception e) {
+      throw Throwables.propagate(e);
+    }
+  }
+
+  /**
+   * Retry copy attempts from input stream to output stream. Does *not* check to make sure data was intact during the transfer
+   *
+   * @param byteSource  Supplier for input streams to copy from. The stream is closed on every retry.
+   * @param byteSink    Supplier for output streams. The stream is closed on every retry.
+   * @param shouldRetry Predicate to determine if the throwable is recoverable for a retry
+   * @param maxAttempts Maximum number of retries before failing
+   */
+  public static long retryCopy(
+      final ByteSource byteSource,
+      final ByteSink byteSink,
+      final Predicate<Throwable> shouldRetry,
+      final int maxAttempts
+  )
+  {
+    try {
+      return RetryUtils.retry(
+          new Callable<Long>()
+          {
+            @Override
+            public Long call() throws Exception
+            {
+              InputStream inputStream = null;
+              OutputStream outputStream = null;
+              try {
+                inputStream = byteSource.openStream();
+                outputStream = byteSink.openStream();
+                return ByteStreams.copy(inputStream, outputStream);
+              }
+              finally {
+                CloseQuietly.close(inputStream);
+                CloseQuietly.close(outputStream);
+              }
+            }
+          },
+          shouldRetry,
+          maxAttempts
+      );
+    }
+    catch (Exception e) {
+      throw Throwables.propagate(e);
+    }
+  }
+
+  public static class MD5MismatchException extends IOException
```

Comment:
maybe something more generic like ChecksumException (like hadoop) would be more useful?